8341273: JVMTI is not properly hiding some continuation related methods

sspitsyn opened the pull request:
This fixes a problem in the VTMS (Virtual Thread Mount State) transition frames hiding mechanism.
Please, see a fix description in the first comment.

Testing:
 - Verified with new test `vthread/CheckHiddenFrames`
 - Mach5 tiers 1-6 are passed

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8341273](https://bugs.openjdk.org/browse/JDK-8341273): JVMTI is not properly hiding some continuation related methods (**Bug** - P3)


### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)
 * [Alex Menkov](https://openjdk.org/census#amenkov) (@alexmenkov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21397/head:pull/21397` \
`$ git checkout pull/21397`

Update a local copy of the PR: \
`$ git checkout pull/21397` \
`$ git pull https://git.openjdk.org/jdk.git pull/21397/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21397`

View PR using the GUI difftool: \
`$ git pr show -t 21397`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21397.diff">https://git.openjdk.org/jdk/pull/21397.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21397#issuecomment-2398039536)

sspitsyn commented:
The frames which are in VTMS transition should not be visible to the JVMTI agents including debug agent because the thread identity can be incorrect. The JVMTI events are not posted when `java_thread->is_in_VTMS_transition() == true`.
All the JVMTI functions returning stack related info do skip frames that are in transition. The hiding mechanism is using the annotation `@JvmtiMountTransition` to mark the `notifyJvmti*` methods and the bit `java_thread->is_in_VTMS_transition()`.
It occurred that the methods `yield()` and `yield0()` can be present in stack trace of an unmounted virtual threads.
The bit `java_thread->is_in_VTMS_transition() ` is not set in such a case.
The fix is to add the annotation `@JvmtiMountTransition` to the methods `yield()` and `yield0()` and correct the frames skipping mechanism to account for such frames as well.
The update also includes:
 - fix in one of the `JvmtiHandshake::execute()` functions which is needed for better stability and safety
 - tweak in the test which expects frames with `yield()` and `yield0()` methods to be present at the top

lmesnik commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -659,33 +661,43 @@ JvmtiEnvBase::check_and_skip_hidden_frames(bool is_in_VTMS_transition, javaVFram
+  bool is_virtual = java_lang_VirtualThread::is_instance(jt->jvmti_vthread());
+
+  if (is_virtual || jt->is_in_VTMS_transition()) { // filter out pure continuations
+    jvf = check_and_skip_hidden_frames(jt->is_in_VTMS_transition(), jvf);
```
Wouldn't be easier to split method `check_and_skip_hidden_frames` to skip_yeilds and skip_transition frames? 
BTW Also it is unclear shouldn't the `@Hidden` methods be skipped from all jvmti frame streams?

sspitsyn replied:
Good suggestion, thanks. I was also thinking about it but was not sure it can be simplified.
JVMTI does not support @Hidden annotation which is used in Java `printStackTrace()` api implementation.

sspitsyn commented on test/hotspot/jtreg/serviceability/jvmti/vthread/CheckHiddenFrames/CheckHiddenFrames.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    public static void main(String[] args) throws Exception {
+        Thread thread = Thread.ofVirtual().unstarted(CheckHiddenFrames::test);
```
Good suggestion, thanks.
Changed to use startVirtualThread.

sspitsyn commented on test/hotspot/jtreg/serviceability/jvmti/vthread/CheckHiddenFrames/CheckHiddenFrames.java:
```diff
@@ -0,0 +1,58 @@
+ */
+
+public class CheckHiddenFrames {
+    private static final String AGENT_LIB = "CheckHiddenFrames";
```
Thanks, removed. I saw it but forgot to remove.

sspitsyn commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -1989,8 +2003,13 @@ void
+  bool is_virtual = java_lang_VirtualThread::is_instance(thread_oop);
 
-  JvmtiVTMSTransitionDisabler disabler(target);
+  // Target can be virtual or platform thread.
```
Thank you for suggestion. Explained comment.
Now, it looks as below:
```
  // Target can be virtual or platform thread.
  // Disable VTMS transition for one thread if it is virtual.
  // Otherwise, disable VTMS transitions for all threads
  // to protect against VTMS transitions in carrier threads.
  // For the latter VTMS transitions are disabled for all threads by several reasons:
  // - carrier threads can mount virtual threads which may cause incorrect behavior
  // - it is good invariant if we can ensure no VTMS transition might happen at some points 
  // - there is no mechanism to disable transitions for a specific carrier threads yet
```

sspitsyn commented:
I've pushed two updates.
The first one addresses the suggestions from Leonid:
 - minor tweaks in new test
 - moved skipping hidden methods `yield()` and `yield0()` from `check_and_skip_hidden_frames()` to a separate function
 - explained in a comment of `JvmtiHandshake::execute()` why all VTMS transitions are disabled for platform thread
 - added extra safety/stability checks and a couple of asserts around calls to `check_and_skip_hidden_frames()`
 
 Second update fixes one more `NotifyFramePop` issue with the frames at the virtual thread start:
  - `enter()`, `enter0`, `VThreadContinuation$1.run()` and `VirtualThread.run()`
  
 This the bottom of stack trace from debugger:
```
  0: java/lang/VirtualThread: run(Ljava/lang/Runnable;)V
  1: java/lang/VirtualThread$VThreadContinuation$1: run()V
  2: jdk/internal/vm/Continuation: enter0()V
  3: jdk/internal/vm/Continuation: enter(Ljdk/internal/vm/Continuation;Z)V
```
The fix includes three parts:
- the `notifyJvmtiStart()`/`notifyJvmtiEnd()` notification calls are moved from `VirtualThread.run()` to the `VThreadContinuation$1.run()`
- the annotation `@JvmtiMountTransition` has been added to the `VThreadContinuation$1.run()`
- the `NotifyFramePop` is changed to return `JVMTI_ERRO_OPAQUE_FRAME` for frames with `enter()` and methods annotated with `@JvmtiMountTransition`

alexmenkov commented on src/hotspot/share/prims/jvmtiExport.cpp:
```diff
@@ -1810,7 +1810,8 @@ void JvmtiExport::post_method_entry(JavaThread *thread, Method* method, frame cu
   }
-  if (mh->jvmti_mount_transition() || thread->is_in_any_VTMS_transition()) {
+  // pure continuations have no VTMS transitions but they use methods annotated with JvmtiMountTransition
+  if ((mh->jvmti_mount_transition() && state->is_virtual()) || thread->is_in_any_VTMS_transition()) {
```
Could you please explain the change (and other similar changes in jvmtiExport.cpp)
Before events were not posted for any `@JvmtiMountTransition` methods, and now only for virtual threads? I.e. events for `@JvmtiMountTransition` methods of carrier threads are posted?

sspitsyn replied:
Good question. Some `jvmtiNotify*` annotations can be called in a context of carrier thread and exited in a context of virtual thread and vice versa. So, you are right we should not post these events for `@JvmtiMountTransition` methods, in a context of carrier threads.
So, we need to adjust these conditions more, something like below:
```
-  if ((mh->jvmti_mount_transition() && state->is_virtual()) || thread->is_in_any_VTMS_transition()) {
+  if ((mh->jvmti_mount_transition() && (state->is_virtual() || thread->last_continuation() == nullptr)) ||
+    thread->is_in_any_VTMS_transition()) {
     return; // no events should be posted if thread is in any VTMS transition
   }
```
The check for `thread->last_continuation() == nullptr` will help to skip these events for carrier threads as well but won't skip them for pure continuations. In fact, I do not like the complexity of this check. :(

sspitsyn commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -674,18 +685,26 @@ JvmtiEnvBase::check_and_skip_hidden_frames(bool is_in_VTMS_transition, javaVFram
+  assert(java_lang_VirtualThread::is_instance(vthread), "sanity check");
+  if (java_lang_VirtualThread::is_instance(vthread)) { // paranoid check for safety
+    if (java_lang_Thread::is_in_VTMS_transition(vthread)) {
+      jvf = check_and_skip_hidden_frames(java_lang_Thread::is_in_VTMS_transition(vthread), jvf);
```
Good catch, thanks. Fixed now.

sspitsyn commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -674,18 +685,26 @@ JvmtiEnvBase::check_and_skip_hidden_frames(bool is_in_VTMS_transition, javaVFram
+
+  if (jt->is_in_VTMS_transition()) {
+    jvf = check_and_skip_hidden_frames(jt->is_in_VTMS_transition(), jvf);
+  } else if (is_virtual) { // filter out pure continuations
```
This seems to be a good catch. The call to `skip_top_jvmti_annotated_frames()` should not be needed.
It does not harm either, so I've added it for safety.
Will remove it and rerun mach5 tiers to make sure there is nothing unexpected.

sspitsyn replied:
Please, hold on. My last comment was not fully incorrect.
Not only `yield()/yield0()` need to be filtered out for virtual threads.
There are more cases for a `JavaThread` which is not in a VTMS transition executing a virtual thread. All the calls below represents inconsistency and must be hidden:

 - `notifyJvmtiStart()`: This function is called in a VTMS MOUNT transition which it ends. So its return happens with the thread identity `VIRTUAL`.
 - `notifyJvmtiEnd()`: This function is called with the thread identity `VIRTUAL`. It is starting a VTMS UNMOUNT transition, so its return happens in a VTMS transition.
 - `notifyJvmtiMount(false)`:  This call happens in a VTMS transition. It finishes the VTMS transition and its return happen with the thread identity 'VIRTUAL'.
 - `notifyJvmtiUnmount(true)`:  This call happens with the thread identity 'VIRTUAL'. It starts a VTMS UNMOUNT transition, so its return happens in a VTMS transition.

I've just discovered that we have two more cases to hide frames for carrier threads that are NOT in a VTMS transitions:
- `notifyJvmtiMount(true)`: This call happens with the thread identity `CARRIER`. It starts a VTMS MOUNT transition, so its return happens in a VTMS transition.
 - `notifyJvmtiUnmount(false)`:  This call happens in a VTMS UNMOUNT transition. It finishes transition, so its return happens with the thread identity `CARRIER`.

An update is needed for these two cases.
All the cases above can be potentially observed at a safepoint/handshake in a non-transitioned state.

sspitsyn commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -1989,8 +2012,17 @@ void
+  oop thread_oop = JNIHandles::resolve_external_guard(target);
+  bool is_virtual = java_lang_VirtualThread::is_instance(thread_oop);
+
+  // Target can be virtual or platform thread.
+  // Disable VTMS transition for one thread if it is virtual.
+  // Otherwise, disable VTMS transitions for all threads.
+  // For the latter, VTMS transitions are disabled for all threads by several reasons:
+  // - carrier threads can mount virtual threads which may cause incorrect behavior
+  // - it is a good invariant when a thread's handshake can't be impacted by a VTMS transition
+  // - there is no mechanism to disable transitions of a specific carrier thread yet
+  JvmtiVTMSTransitionDisabler disabler(is_virtual ? target : nullptr); // nullptr is to disable all
```
Good suggestion, thanks.
Though, I'd prefer to fix it here, so will prepare a fix in the `JvmtiVTMSTransitionDisabler` constructor.

AlanBateman commented on src/java.base/share/classes/java/lang/VirtualThread.java:
```diff
@@ -213,8 +213,15 @@ protected void onPinned(Continuation.Pinned reason) {
+                    // notify JVMTI
+                    vthread.notifyJvmtiStart();
+
                     vthread.run(task);
```
This doesn't look right, it needs to use try-finally.

sspitsyn replied:
The discussion on this is on the next comment.

AlanBateman commented on src/java.base/share/classes/java/lang/VirtualThread.java:
```diff
@@ -417,8 +421,6 @@ private void run(Runnable task) {
                 }
 
             } finally {
```
This means an empty finally block, I assume you'll remove the try-finally here.

sspitsyn replied:
Thank you for the comment. I can move the try-finally to the method `java/lang/VirtualThread$VThreadContinuation$1.run()` if you prefer. But it will play the same role functionally.

AlanBateman replied:
Changing the run method to
```
vthread.notifyJvmtiStart();
try {
    vthread.run(task);
} finally {
    vthread.notifyJvmtiEnd();
}
```
is okay. I was initially puzzled because the run method is already`@Hidden` but I don't think this is used by JVMTI.

sspitsyn replied:
Thanks. Yes, the annotation `@Hidden` is not used by the JVMTI. It would be tricky to use it consistently.

AlanBateman commented on src/java.base/share/classes/java/lang/VirtualThread.java:
```diff
@@ -213,8 +213,15 @@ protected void onPinned(Continuation.Pinned reason) {
+                @JvmtiMountTransition
                 public void run() {
+                    // notify JVMTI
+                    vthread.notifyJvmtiStart();
```
The notifyJvmtiMount and notifyJvmtiUnmount (native) methods already have the annotations. Is it really required on the caller too?  I'm wondering if the comment on JvmtiMountTransition needs to be expanded to explain this.

sspitsyn replied:
The method `java/lang/VirtualThread$VThreadContinuation$1.run()` is starting and finishing in a VTMS transition.
The issue is with the JVMTI `NotifyFramePop`. We need a way to disallow adding `FramePop` event requests to its frame. I'm trying to move the `notifyJvmtiStart/notifyJvmtiEnd` calls to earlier frame to reduce a little bit the scope of VTMS transition. What is the best place to explain it?

AlanBateman replied:
No issue from me on moving the notifications to the run method. My comment was more for the JvmtiMountTransition's class description as it's not easy to audit the use of this annotation.

sspitsyn replied:
Got it, thanks.

AlanBateman replied:
The updated changes to VirtualThread look fine. I'm still wondering about JvmtiMountTransition's class description.  There's nothing obvious in yield and yield0 that they start/end a VTMS.  Maybe for a future PR but I think the description could be expanded as the same questions will come up each time that anything using this annotation changes.

sspitsyn commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -674,18 +685,26 @@ JvmtiEnvBase::check_and_skip_hidden_frames(bool is_in_VTMS_transition, javaVFram
+  bool is_virtual = java_lang_VirtualThread::is_instance(jt->jvmti_vthread());
+
+  if (jt->is_in_VTMS_transition()) {
+    jvf = check_and_skip_hidden_frames(jt->is_in_VTMS_transition(), jvf);
```
Fixed, thanks.

sspitsyn commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -583,8 +583,10 @@ JvmtiEnvBase::jvf_for_thread_and_depth(JavaThread* java_thread, jint depth) {
-
+  // There should not be any VTMS transition here. This is for safety.
+  if (java_thread->is_in_VTMS_transition()) {
+    jvf = JvmtiEnvBase::check_and_skip_hidden_frames(java_thread, jvf);
```
Thank you for the comment. Unfortunately, with my latest update it is not relevant anymore.

sspitsyn commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -725,9 +745,11 @@ JvmtiEnvBase::get_cthread_last_java_vframe(JavaThread* jt, RegisterMap* reg_map_
   if (jt->is_in_VTMS_transition()) {
+    // Skip hidden frames only for carrier threads
+    // which are in non-temporary VTMS transition.
     jvf = check_and_skip_hidden_frames(jt, jvf);
```
Thank you for the comment. Unfortunately, with my latest update it is not relevant anymore.

sspitsyn commented:
I believe, I've resolved almost all review comments.
Please, let me know if anything is missed or there are some issues in my latest updates.
The mach5 tiers 1-6 are good.

AlanBateman commented on src/java.base/share/classes/jdk/internal/vm/annotation/JvmtiMountTransition.java:
```diff
@@ -30,9 +30,12 @@
  * or ends virtual thread mount state transition (VTMS transition).
+ * The Continuation yield and yield0 frames normally are in VTMS transition
+ * but can be found out of transition in an unmounted virtual thread.
+ * This inconsistency is the reason why they also need this annotation.
```
In ChangesCurrentThread, Hidden and the other annotations, the description explains where they must be used. The JvmtiMountTransition description isn't very clear as it's explain why the Continuation yield methods need the annotation. It's okay if the description is re-worked in some future PR, my comment is just that the current description is very confusing.

AlanBateman commented on src/java.base/share/classes/jdk/internal/vm/annotation/JvmtiMountTransition.java:
```diff
@@ -28,11 +28,18 @@
- * A method is annotated as "jvmti mount transition" if it starts
- * or ends virtual thread mount state transition (VTMS transition).
+ * A method may be annotated as "jvmti mount transition" to hint
+ * it is desirable to omit it from JVMTI stack traces.
```
Might be better to replace both usages of "jvmti mount transition" with JvmtiMountTransition.

"virtual thread mount state transition (VTMS transition)" should probably be "Virtual Thread Mount State (VTMS) transition".

The updated wording is better but I think this still hard to audit to know if you've got the usages in the right place. Maybe we can re-visit it in the future.

sspitsyn replied:
I believe it has been fixed with the latest updates.

AlanBateman commented:
Latest update to JvmtiMountTransition looks okay. I assume wait until Alex Menkov and Leonid are finished their reviews.

alexmenkov commented on src/java.base/share/classes/jdk/internal/vm/annotation/JvmtiMountTransition.java:
```diff
@@ -28,11 +28,14 @@
  *
  * @implNote
- * This annotation is only used for VirtualThread methods.
+ * This annotation is only used for the VirtualThread notifyJvmti*  methods.
```
What about VirtualThread.switchToCarrierThread and VirtualThread.switchToVirtualThread ? They also have the annotation.

AlanBateman replied:
We working to remove these two methods. I think the main thing for JvmtiMountTransition is that the interface description provides enough information to know when the annotation is needed.

sspitsyn replied:
Good catch, thanks. These two methods are impacted by temporary VTMS transitions.
I've restored as it was before.

alexmenkov commented on src/java.base/share/classes/jdk/internal/vm/annotation/JvmtiHideEvents.java:
```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (c) 2021, 2022, Oracle and/or its affiliates. All rights reserved.
```
(c) 2024

sspitsyn replied:
Fixed, thanks.

sspitsyn commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -652,22 +651,31 @@ JavaThread* JvmtiEnvBase::get_JavaThread_or_null(oop vthread) {
+}
+
 javaVFrame*
 JvmtiEnvBase::check_and_skip_hidden_frames(bool is_in_VTMS_transition, javaVFrame* jvf) {
```
Good catch, thanks.
These two functions are impacted by temporary VTMS transitions. It seems we fail to hide frames in these transitions while we skip the JVMTI events in their context. I'll try to fix this issue.

sspitsyn replied:
I'd suggest to file a separate bug on this issue as it has to be tested well.

sspitsyn replied:
Alan has a plan and fix to get rid of temporary transitions from the `VirtualThread` class.

AlanBateman commented on src/java.base/share/classes/jdk/internal/vm/annotation/JvmtiHideEvents.java:
```diff
@@ -0,0 +1,40 @@
+
+/**
+ * A method may be annotated with JvmtiHideEvents to hint it is not
+ * desirable to sent JVMTI events in context of the annotated method.
```
s/sent/send/ 

Also might be cleaerr to drop "not desirable" from the sentence as it begs too many questions.

sspitsyn replied:
Thanks, fixed now.

AlanBateman commented on src/java.base/share/classes/jdk/internal/vm/annotation/JvmtiHideEvents.java:
```diff
@@ -0,0 +1,40 @@
+
+/**
+ * A method may be annotated with JvmtiHideEvents to hint the JVMTI events
+ * should not be generated in context of the annotated method.
```
The sentence isn't quite right. I think you want to say that it's "a hint that JVMTI events" or "to hint that JVMTI events".

sspitsyn commented:
Alan and Alex, thank you for review!

sspitsyn commented:
/integrate
